### PR TITLE
[widget Audit] toga.Canvas

### LINKED ISSUE #730: Canvas documentation error? (no attribute 'translate')
## Expected Behavior

I'm trying to run the code samples from the [Canvas documentation](https://toga.readthedocs.io/en/latest/reference/api/widgets/canvas.html). They should run flawlessly.

## Current Behavior

Code sample 3 ("example with a platformer game") fails with an `AttributeError`:

```python
AttributeError: 'Context' object has no attribute 'translate'
```
The affected line is:
```python
hero.translate(10, 0)
```

## Steps to reproduce

1. Generate a briefcase cross-platform app
2. Add code sample 3 from the Toga Canvas docs
3. Run `python3 setup.py linux -s`

## Your Environment

* Python Version: 3.6.8

* Operating System and Version:

  - [ ] macOS - version:
  - [x] Linux - distro: Ubuntu  - version: 18.04.3 LTS Bionic
  - [ ] Windows - version:
  - [ ] Other - name:  - version:

* Toga Version: 0.3.0.dev12

* Toga Target (the type of app you are trying to generate)

  - [ ] android
  - [ ] cocoa
  - [ ] django
  - [x] gtk
  - [ ] iOS
  - [ ] tvOS
  - [ ] watchOS
  - [ ] winforms
  - [ ] win32
  - [ ] Other (please specify)

### LINKED ISSUE #996: Drawing from tutoral 4 does not match the screenshot
**Describe the bug**
The drawing doesn't match the picture. 😉 

**To Reproduce**
1. Go to https://toga.readthedocs.io/en/latest/tutorial/tutorial-4.html
2. Run it
3. Look at the picture and see how it looks like a [CSS joke mug](https://www.zazzle.com/rlv/css_is_awesome_coffee_mug-r2b2e63ffaa0a46628fb84f0844e7bf9d_x7jg9_8byvr_512.jpg)

**Expected behavior**
Match the picture.

**Screenshots**
image

**Environment:**
 - Operating System: Windows 10
 - Python version:
 - Software versions: Cpython 3.8.3
   - Briefcase:
   - Toga: 0.3.0dev22
   - ...

### LINKED ISSUE #1713: Simplify API for canvas event handlers
### What is the problem or limitation you are having?

The current implementation of event handlers on toga.Canvas has some interesting quirks.

1. It allows for a "double click drag"... and I'm not even sure what that would mean.
2. It is closely bound to a mouse-driven implementation (in particular, it references alt and double click)

### Describe the solution you'd like

It might make more sense to drop the `clicks` argument to the event handlers, and have separate events for:
* on_press and on_alt_press
* on_activate (i.e., double click) and on_alt_activate
* on_drag
* on_release and on_alt_release

That would make events a lot closer to their "functional" purpose, rather than tying them to a specific implementation (e.g. - what is a double click on a touchscreen?) It would de-emphasize the importance of clicks, especially in the context of drag; and it would put the focus on press/activate, rather than release events.

The general advice would be to avoid using "alt" press (or, at least, provide different ways of accessing that functionality); this is consistent with the sort of approach taken by the macOS style guide, which allows for "right click" actions, but encourages that to be an interface for "power users".

This would simplify the winforms and GTK implementations (which currently need to do imml-level stateful tracking of the current click state), and makes the API for mobile canvas less of a band-aid (as we can use long-press or force-press events for "activate").

### Describe alternatives you've considered

Do nothing. The current API works; it's just not ideal.

### Additional context

_No response_

### LINKED ISSUE #1991: GTK Canvas doesn't set font description correctly
### Describe the bug

Canvas crashes with `TypeError` on Linux.

```
Traceback (most recent call last):
  File "/home/russell/github/beeware/toga/examples/canvas/build/canvas/pop/jammy/canvas-0.0.1/usr/lib/canvas/app_packages/toga_gtk/app.py", line 98, in gtk_startup
    self.interface.startup()
  File "/home/russell/github/beeware/toga/examples/canvas/build/canvas/pop/jammy/canvas-0.0.1/usr/lib/canvas/app/canvas/app.py", line 158, in startup
    self.render_drawing()
  File "/home/russell/github/beeware/toga/examples/canvas/build/canvas/pop/jammy/canvas-0.0.1/usr/lib/canvas/app/canvas/app.py", line 378, in render_drawing
    self.draw_shape(context)
  File "/home/russell/github/beeware/toga/examples/canvas/build/canvas/pop/jammy/canvas-0.0.1/usr/lib/canvas/app/canvas/app.py", line 388, in draw_shape
    drawing_instructions(context, factor)
  File "/home/russell/github/beeware/toga/examples/canvas/build/canvas/pop/jammy/canvas-0.0.1/usr/lib/canvas/app/canvas/app.py", line 509, in draw_instructions
    width, height = self.canvas.measure_text(text, font, tight=True)
  File "/home/russell/github/beeware/toga/examples/canvas/build/canvas/pop/jammy/canvas-0.0.1/usr/lib/canvas/app_packages/toga/widgets/canvas.py", line 748, in measure_text
    return self._impl.measure_text(text, font, tight=tight)
  File "/home/russell/github/beeware/toga/examples/canvas/build/canvas/pop/jammy/canvas-0.0.1/usr/lib/canvas/app_packages/toga_gtk/widgets/canvas.py", line 232, in measure_text
    layout.set_font_description(self.native)
TypeError: argument desc: Expected Pango.FontDescription, but got gi.repository.Gtk.DrawingArea
```

### Steps to reproduce

Run `briefcase run linux system` for the `canvas` or `tutorial4` examples.

### Expected behavior

Does not crash.

### Screenshots

_No response_

### Environment

- Operating System: pop os 22.04
- Python version: 3.10.12
- Software versions:
  - Briefcase: `0.3.15.dev361+g02f412be`
  - Toga: `toga-gtk-0.3.2.dev380+g7408a792.d20230618`

### Logs

[briefcase.2023_06_18-16_23_56.run.log](https://github.com/beeware/toga/files/11782517/briefcase.2023_06_18-16_23_56.run.log)

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by freakboy3742**
Audit of Canvas.

Builds on #1903 as an audit of Font.

This involves a number of backwards incompatible changes: 
* Canvas is no longer its own context. This previously allowed the use of simple drawing operations on the canvas itself, which is fine; but the context also allows for *manipulation* of primitive drawing operations using `remove` and `clear`, which collide with Widget's child manipulation operations of the same name. Interestingly, the method to add was called `add_draw_obj`, and was explicitly undocumented. To avoid the collision, Canvas now has a `context` property that returns the root context of the canvas. Simple drawing operations have been retained for backwards compatibility, redirecting to the default context; context generating operations have been retained.
* To make it clear(er) when an operation is a context producing operation, rather than a simple drawing operation, context-producing operations have been renamed to TitleCase. This also avoids a collision caused by introducing the `Canvas.context` property; the operation on canvas to produce a new sub-context is now `Canvas.Context()`.
* The `preserve` option on a `Fill()` context has been removed. It appears to be an internal optimization for GTK that isn't required, and isn't easy to explain.
* The `new_path` simple drawing operation has been renamed `begin_path` for consistency with HTML5.
* The "clicks" argument to event handlers has been removed, in favor of a "on_activate" handler.
* On GTK, font and stroke handling isn't quite right. It works (to first approximation). This seems to be deeply seated in how fonts are rendered by Cairo; and there's a testing issue with consistent availability of fonts.

Fixes:
* Fixes #730.
* Fixes #996.
* Fixes #1713.
* Fixes #1991.

## Audit checklist
* [x]  Core tests ported to pytest
* [x]  Core tests at 100% coverage
* [x]  Core docstrings complete, and in Sphinx format
* [x]  Documentation complete and consistently formatted
* [x]  cocoa backend at 100% coverage
* [x]  winforms backend at 100% coverage
* [x]  gtk backend at 100% coverage
* [x]  iOS backend at 100% coverage
* [x]  android backend at 100% coverage
* [x]  Widget support table updated
* [x]  Widget Issue backlog reviewed

### CODE REVIEW

**Comment by freakboy3742** [comment]
So I don't forget where I'm up to: the test failures at this point are all due to weird inconsistencies between CI behavior and desktop behavior.

Linux:
* multiline text: Desktop multiline_text-linux CI multiline_text-linux

iOS:
* multiline text: Desktop multiline_text-iOS CI multiline_text-iOS

macOS:

* multiline text: Desktop multiline_text-macOS CI: 
multiline_text-macOS

* transparency: Desktop: transparency CI: transparency-macOS

The differences on multiline text all seem to be related to line leading.

The transparency differences are due to the alpha calculation on the transparent rectangle.

**Comment by freakboy3742** [comment]
@mhsmith Ok - after further investigation, I think we're stuck between a rock and a hard place.

In the multiline text case, AFAICT the discrepancies are caused by font leading, which is caused by subtle differences in the font itself between OS versions. I'm not sure I see any way to work around this; and in the Linux case, it's going to be *hugely* dependent on what the local system, GTK theme, and more.

In the transparency case, I can reproduce the problem moving between my M1 MacBook and my x86_64 Mac mini. The issue appears to be that recent MacBooks have [a different display profile](https://support.apple.com/en-au/guide/mac-help/mchlf3ddc60d/mac), so they operate in sRGB colourspace; older Macs don't do this by default - although the colourspace is extremely configurable. 

So - the options I see are:

1. Have multiple reference images. This will let us pass CI, but it will be difficult to guarantee a CI configuration that will pass for *every* setup.
2. Use a weaker measure of similarity for the test. For example, in the multiline text case, we could count the number of white pixels in the image, and ensure that number is in a given bound; and maybe refer to individual pixel checking to ensure that some locations are the right color (picking a location like the middle of an L upright as the target location). This will make it much easier for a problem to slip in, as the conditions for "passing" the test are much looser.

Any preferences or alternatives that I may have missed?

**Comment by freakboy3742** [comment]
Actually - thinking about this some more, there's another option (for macOS and iOS anyway) - we can make the variants dependent on the macOS/iOS version. 

The same approach might work for GTK; but we'll need more variants. However, it's a moot point at present because GTK font rendering has deeper problems.

**Comment by freakboy3742** [comment]
After many hours of staring intensely at image diffs, I've come to the conclusion is that the only fix is bigger error tolerances. 

In the transparency test, the errors are entirely around the edge of the image - subpixel antialiasing discrepancies introduced by a combination of image scaling and alpha blending. I can't find any way to ignore those errors that won't also cause a "translate the image by 1 pixel" error to inadvertently pass. Increasing the error threshold allows both to pass. 

The same is true of the font tests. There's a per-platform font discrepancy which is easy enough to allow for; but font rendering is the worst case for subpixel antialiasing issues, because they're *entirely* edges. 

So - I'm going to say that the best we can do is to loosen the error thresholds enough that they pass on CI, and just accept that any changes to canvas rendering will involve some manual inspection as well as automated testing... which was probably true regardless.

File: docs/reference/api/widgets/canvas.rst [hunk 1]
```diff
@@ -105,21 +105,52 @@ Notes
...
-.. autoclass:: toga.widgets.canvas.Context
-
-Simple Drawing objects
-^^^^^^^^^^^^^^^^^^^^^^
-
-.. automodule:: toga.widgets.canvas
-   :exclude-members: Canvas, Context, ClosedPathContext, FillContext, StrokeContext
+    :exclude-members: new_path, move_to, line_to, bezier_curve_to, quadratic_curve_to,
+        arc, ellipse, rect, write_text, rotate, scale, translate, reset_transform,
+        closed_path, fill, stroke
```
Comment (mhsmith) [inline docs/reference/api/widgets/canvas.rst:114]: I know we normally include deprecated methods in the documentation, but in this case there are so many of them that they would make it impossible to see an overview of the non-deprecated Canvas API. It's not critical to document them, because anyone who's using them will get a DeprecationWarning telling them exactly what to do.

**Comment by mhsmith** [comment]
> On GTK, font and stroke handling isn't quite right. It works (to first approximation). This seems to be deeply seated in how fonts are rendered by Cairo; and there's a testing issue with consistent availability of fonts.

What's the current status of this, and could it be fixed by using a font file rather than a system font?

File: changes/2029.removal.8.rst [hunk 2]
```diff
@@ -1 +1 @@
-The ``tight`` argument has been removed from ``Canvas.measure_text()``. It can be removed without impact.
+The ``tight`` argument to ``Canvas.measure_text()`` has been deprecated. It was never implemented, and can be safely removed without impact.
```
Comment (freakboy3742) [inline changes/2029.removal.8.rst:1]: Not entirely true - it was implemented on GTK (surfacing a platform implementation detail).

**Review by freakboy3742 — COMMENTED** [review]
The updates and changes here all make sense.

I did notice a small inconsistency in the handling of quadratic curves:

Screenshot 2023-09-18 at 8 15 27 am

I believe this is because it's drawing a cardinal spline, rather than a quadratic curve. I had a similar issue on GTK; the solution is to lean on the fact that a "Bézier" is actually a "Bézier cubic", and a quadratic curve is just a dimensionally reduced Bézier cubic. 

It's also not 100% clear to me if the larger font size is in the bounds of "expected inconsistency", or if the DPI scaling is off in this case.

File: core/src/toga/widgets/base.py [hunk 3]
```diff
@@ -167,8 +177,9 @@ def remove(self, *children: Widget) -> None:
         :param children: The child nodes to remove.
         :raises ValueError: If this widget cannot have children.
         """
-        removed = False
+        self._assert_can_have_children()
 
+        removed = False
```
[the remark this answers is not in the document]
Comment (mhsmith) [reply core/src/toga/widgets/base.py:178]: I didn't remove the line, I only adjusted the spacing.

File: changes/2029.removal.8.rst (hunk 2 above)
Comment (mhsmith) [reply changes/2029.removal.8.rst:1]: OK, I'll reword it to match the note on `preserve`.

File: core/src/toga/widgets/canvas.py [hunk 4]
```diff
@@ -23,7 +23,34 @@
...
+    Every context drawing method creates a ``DrawingObject``, adds it to the context,
+    and returns it. Each argument passed to the method becomes a property of the
+    ``DrawingObject``, which can be modified as shown in the `Usage`_ section.
+
+    ``DrawingObjects`` can also be created manually, then added to a context using the
+    :meth:`~Context.append` or :meth:`~Context.insert` methods. Their constructors take
+    the same arguments as the corresponding :any:`Context` method, and their classes
+    have the same names, but capitalized:
+
+    * ``toga.widgets.canvas.Arc``
```
Comment (freakboy3742) [inline core/src/toga/widgets/canvas.py:37]: Maybe these should be `:meth:` links, rather than static strings?

File: cocoa/src/toga_cocoa/fonts.py [hunk 5]
```diff
@@ -55,11 +55,13 @@ def __init__(self, interface):
...
 
-            # Default system font size on Cocoa is 12pt
             if self.interface.size == SYSTEM_DEFAULT_FONT_SIZE:
                 font_size = NSFont.systemFontSize
             else:
-                font_size = self.interface.size
+                # A "point" in Apple APIs is equivalent to a CSS pixel, but the Toga
+                # public API works in CSS points, which are slightly larger
+                # (https://developer.apple.com/library/archive/documentation/GraphicsAnimation/Conceptual/HighResolutionOSX/Explained/Explained.html).
+                font_size = self.interface.size * 96 / 72
```
Comment (freakboy3742) [inline cocoa/src/toga_cocoa/fonts.py:64]: This interpretation of font size makes sense - but it seems like something we should capture on the Font documentation page, along with platform notes for each platform that provides specifics for the benefit of anyone who might have platform experience.

Those details should then be cross-linked to the Pack style page discussing `font-size` attribute.

**Comment by mhsmith** [comment]
>  I had a similar issue on GTK; the solution is to lean on the fact that a "Bézier" is actually a "Bézier cubic", and a quadratic curve is just a dimensionally reduced Bézier cubic.

Done.

> It's also not 100% clear to me if the larger font size is in the bounds of "expected inconsistency", or if the DPI scaling is off in this case.

In this case it's not caused by DPI scaling, but by differences in the definition of a "point". Windows uses the CSS definition, while Apple makes a point equal to a logical pixel. Fixed by increasing font sizes on Apple, as mentioned above.

**Comment by mhsmith** [comment]
Android does have a "point" unit in its API, but it's not generally used even for font sizes, and doesn't have a clearly documented definition, so I'll use 1.33 logical pixels there as well.

File: testbed/src/testbed/resources/canvas/write_text-macOS.png
Comment (mhsmith) [inline testbed/src/testbed/resources/canvas/write_text-macOS.png:1]: I can only explain the overlapping stroke on the "a" as being a bug in the font itself. But when we implement font loading support on Cocoa, this file will be replaced anyway.

**Comment by mhsmith** [comment]
I've changed the default WinForms font to respect the system theme, a change which has already been made in newer versions of .NET (https://github.com/dotnet/winforms/pull/656), though as I understand it, not in the version of .NET which comes with Windows itself.

The difference is not dramatic, but readability is definitely improved: see screenshots at https://github.com/dotnet/winforms/pull/656#issuecomment-499766882. Notice that the title bar was already using the system theme, and the menu bar was too.

The concerns in that comment don't apply to us because we're not using "hand placed controls with pixel perfect layout".

File: cocoa/src/toga_cocoa/fonts.py (hunk 5 above)
Comment (mhsmith) [reply cocoa/src/toga_cocoa/fonts.py:64]: Done.

File: docs/background/topics/layout.rst [hunk 6]
```diff
@@ -26,9 +26,7 @@ When a refresh is requested on a container, any sub-containers will also be refr
 Length units
 ============
 
-Toga uses CSS units in its public API. All size calculations assume that the user has a
-96 DPI display; if the user has a higher density display, the sizes that are applied
-will be scaled accordingly. In practice, this means that:
+Toga uses CSS units in its public API. Their physical size depends on the device type:
```
Comment (mhsmith) [inline docs/background/topics/layout.rst:30]: I think privileging 96 DPI in this context causes more confusion than it solves. If the user wants a full explanation, they can always follow the link.

**Review by mhsmith — APPROVED** [review]

File: testbed/src/testbed/resources/canvas/arc.png
Comment (freakboy3742) [inline testbed/src/testbed/resources/canvas/arc.png:1]: Art immitates life... the test case now reflects the what how *we* feel about the Canvas widget :-)
